### ISSUE TITLE
Unable to remove 'charset' field from 'Content-Type' of text types - HTTP

### ISSUE THREAD

**Issue by gankisreal** [symfony/symfony#20357]
Hey guys. Client expects `Content-Type: text/plain` header in my responses, `Response::setCharset` and `RequestBag::remove` doesnt allow me to achieve that. More precisely some step in symfony kernel must be calling `prepare()` on my response which stuffs `charset=UTF-8` into my response. And that unfortunately breaks this client. Client breaks on `charset=US-ASCII` either. Response filters does not remove it either as supposedly should (as a last resort). Does that deserve fix?
Symfony 3.1.5

**Comment by patrick-mcdougle** [comment]
The [IETF RFC 2046](https://tools.ietf.org/html/rfc2046#section-4.1.2) uses the word 'MAY' and not 'MUST', so I think this is a valid complaint/bug.

**Comment by gankisreal** [comment]
> The IETF RFC 2046 uses the word 'MAY' and not 'MUST', so I think this is a valid complaint/bug.

Yes, that's why any existing or future sub-types of text may not use 'charset' at all, then why to send it? It's even more prominent for types that follow [RFC6657, section 3](https://tools.ietf.org/html/rfc6657#section-3) option (a).

**Some other implications**:
For types that require explicit inclusion of the "charset": it shouldn't be the framework's responsibility to include "guessed" value, but developers/software's he's producing. If for such type bundle dont produce 'charset' field, client can recognize error. If conventionaly UTF-8 is stuffed client might recognize it as valid response and start to process it in a way it shouldnt. That has similar effect to client side content sniffing for character-set recognition described in RFC's security sections as bad.

Another bad case I faced myself is with HTML that use character-set other than UTF-8. 'charset' in [text/html](https://www.iana.org/assignments/media-types/text/html) is specified as:
> may be provided to definitively specify the document's character encoding, overriding any character encoding declarations in the document

Thus for every document with `meta charset=UTF-16` (or what not) developer is forced to invoke `Response::setCharset` increasing error chance.

Extended version of above case is when any text sub-type defined similarly to 'text/html', which content is unknown (eg. external sourced documents), require developer to parse document for character-set declaration extraction, which then he can apply to response.

Similar case with types from previous case. If character-set of such document is unknown to dev and there is no knowledge/resources to parse character-set declaration from such documents, might lead to sending invalid response.

The RFC6657 is a strong call for 'text' types to unambigously specify how character-set is served. Symfony's support for these is, in my opinion, insufficient.

**Comment by gankisreal** [comment]
Also my issue with client not accepting `Content-Type` other than `text/plain` is a call for improvement in a framework for not HTTP compilant clients. My client is pretty bad, I wont lie. But I'm forced to work with it. Magic in framework is/might be useful, and it has it's place in HttpKernel pipeline (controller fragment). Why not leverage pipeline and allow to hook a code that modify HTTP headers to uncompilant state. Magic can take place in controller but if someone is hooking filter with negative priority and modify headers to uncompilant state there, presumably clearly knows what he's doing. Or if someone uses `CustomResponse` documented as "no magic inside" is presumably knows what he's doing. Might be he's working with semi-HTTP-compilant client.

**Comment by gankisreal** [comment]
This can be closed, made an error duo to misinterpretation of php documentaiton regarding 'default_charset' ini option. 
**To solve:**
> default_charset -- All versions of PHP will use this value as the charset within the default Content-Type header sent by PHP if the header isn't overridden by a call to header(). 

Calling `header()` isnt enough to override 'charset' field call to `ini_set('default_charset', '')` does the job. Then filter with priority -1 is enough to have control over 'charset'. Or extending and overriding `Response::prepare()` is enough. 
Cumbersome but I'll take it. Sorry for disturbance.

**Comment by alecpl** [comment]
It shouldn't have been closed. Obviously the charset= is being added by Symfony's Response.

https://github.com/symfony/http-foundation/blob/74d171d5b6a1d9e4bfee09a41937c17a7536acfa/Response.php#L267

I just got hit with this issue when testing KTOR's SSE plugin. It does not expect charset in the Content-Type and throws exception. While it is an obvious KTOR issue (that got fixed already in unreleased version) we obviously need a way to send the header without charset. The content-type in my case is `text/event-stream`.

I haven't found a workaround.
